WIP: Improve IO tutorial

cgohlke opened the pull request:
## Description

This PR improves the `phasorpy_io.py` tutorial. So far:

- Use "phase histogram" instead of "TCSPC histogram" to describe the signal returned by the `signal_from_fbd` function. That is more in-line with FLIMbox terminology.
- Add an option to label the x-axis of the averaged signal plot in the `plot_signal_image` function.

## Checklist

- [ ] The pull request title and description are concise.
- [ ] Related issues are linked in the description.
- [ ] New dependencies are explained.
- [ ] The source code and documentation can be distributed under the [MIT license](https://www.phasorpy.org/docs/stable/license/).
- [ ] The source code adheres to [code standards](https://www.phasorpy.org/docs/stable/contributing/#code-standards).
- [ ] New classes, functions, and features are thoroughly [tested](https://www.phasorpy.org/docs/stable/contributing/#tests).
- [ ] New, user-facing classes, functions, and features are [documented](https://www.phasorpy.org/docs/stable/contributing/#documentation).
- [ ] New features are covered in tutorials.
- [ ] No files other than source code, documentation, and project settings are added to the repository.

cgohlke commented:
Hi @bruno-pannunzio , when you have the time, can you add your comments for the IO tutorial to this PR? 

Items to discuss:

- Did you have a chance to export a PTU file from LIF using the photon filter ([as discussed here](https://github.com/phasorpy/phasorpy/pull/188#pullrequestreview-2618569009))?

- While FLIMbox terminology uses "phase histogram", FLIM LABS (based on another implementation of the FLIMbox AFAIK) uses "TCSPC histogram". See [here](https://www.flimlabs.com/techniques/tcspc/) for example.

bruno-pannunzio commented:
Hi @cgohlke! I am reviewing it and will come back with all the comments. In the meantime, I have the LIF file for HSI Maria (from i3S) sent if you want to include it in the tutorial. I think you are also cc'd in the email. I opened the file and it works great with the `signal_from_lif` function and seems to have what is needed for the tutorial.

> * Did you have a chance to export a PTU file from LIF using the photon filter ([as discussed here](https://github.com/phasorpy/phasorpy/pull/188#pullrequestreview-2618569009))?

We don't have acces yet to the LAS X software. I just sent a remainder to see if we can get access this week. Alternatively we can also ask Maria if they can open it if we see the access for the software is going to take a while.

> * While FLIMbox terminology uses "phase histogram", FLIM LABS (based on another implementation of the FLIMbox AFAIK) uses "TCSPC histogram". See [here](https://www.flimlabs.com/techniques/tcspc/) for example.

I agree with Leonel comments regarding the TCSPC. I don't know exactly the technology behind the FLIM LABS hardware but we can ask Alessandro. But since the FLIMbox works with the Digital Frequency Domain principle, I agree that TCSPC would not be a correct terminology in that case. Also I am not sure all FLIMbox work with the DFD principle, maybe we can ask @lmalacrida.

cgohlke commented:
> I have the LIF file for HSI Maria (from i3S) sent if you want to include it in the tutorial

Yes, if the file can be shared on Zenodo, I can integrate it in this PR.

>  But since the FLIMbox works with the Digital Frequency Domain principle, I agree that TCSPC would not be a correct terminology in that case

I'm not that sure. The argument was about "TCSPC histogram" specifically. I think the original purpose of the FLIMbox was to obtain the equivalent of TCSPC histograms for phasor analysis in a more direct, efficient, and less error prone way than existing hardware and software at that time. Ultimately there should be no difference between the histograms from time-domain and digital frequency-domain measurements for phasor analysis. Also the techniques are not that different: both are counting single photons, are based on correlation, and are using time-tagged records. Maybe we can use "CSPC histogram". Just kidding. Let's go with FLIMbox terminology for FBD.

bruno-pannunzio commented:
> if the file can be shared on Zenodo,

Yes! Maria already gave us the OK to publish it in Zenodo so we can go ahead.

> Ultimately there should be no difference between the histograms from time-domain and digital frequency-domain measurements for phasor analysis.

You are totally right! This concept is something we are trying to push more into the community, particularly in our recent review. I think in concept is the same, but also I think it's good to differentiate how the acquisition was made. 

Regarding the tutorial, I agree with what you said in the last meeting that practically al files can be improved, since all have something 'weird'. But at least the basic workflow for reading the files is very clear, which is the main purpose of the tutorial I think. I will add some more comments in line.

bruno-pannunzio commented on tutorials/api/phasorpy_io.py:
```diff
@@ -267,7 +268,7 @@
+plot_signal_image(signal, title=filename, xlabel='wavelength (nm)')
 
 # %%
 # Note that the signal is not well sampled and shows discontinuity.
```
I like this comment, although the file is not optimal, it is also useful to raise awareness of this concept

cgohlke replied:
As far as I understand, the phasor coordinates obtained by DFT on this kind of signals may not be a good approximation of the real phasor coordinates (which are defined as integrals mathematically). Anyway, we might want to add an option to use better integration methods (from `scipy.integrate`?). The large discontinuity at the higher wavelength end seems worrisome regardless. But maybe all of this is still "good enough"? Has this been investigated before for hyperspectral imaging?

lmalacrida replied:
Yes, @cgohlke, we studied the discontinuity issue at the higher wavelength in this paper's supp material ( doi: 10.1038/s42003-021-02266-z.). It's not a big issue.

cgohlke replied:
Good to know. Interesting, since it's such a big issue for lifetime data. But that field is more quantitative.

cgohlke replied:
Hi @bruno-pannunzio given @lmalacrida's comment, should the note be removed?

bruno-pannunzio replied:
I believe we can remove it since is also not strictly related to IO of files. We can then create a separate tutorial for HSI analysis and it will be more appropriate to mention it there. What do you think?

bruno-pannunzio commented:
Some more comments related to the tutorial:

1. Regarding the LIF file position in the phasor plot. I don't know if it is correct because I am not aware which sample is and how it was acquired. Also, since we don't know how the reference phase and modulation values for calibration were calculated, I don't know if that can also add some discrepancies to the location in the phasor. I know they perform calibration with simulated IRF, so maybe if this is related to how those values were calculated, it can give incorrect location when using our calibration. 

2. The differences in the lifetime histogram between the PTU and LIF are interesting, and I think we can get more information when we access the LAS X software. But in the meantime, since the tutorial is about reading and writing files, I don't know if this is strictly related to the topic of the tutorial and may be omitted?

3. I believe the location of the FBD file of the convallaria is correct, at least is what we see in the acquisition software.

cgohlke commented:
> Yes! Maria already gave us the OK to publish it in Zenodo so we can go ahead.

Can you submit the file to the [PhasorPy community on Zenodo](https://zenodo.org/communities/phasorpy/records)? Give proper credit to Maria and add a short description of the sample and relevant information about instrumentation/acquisition parameters.

> The differences in the lifetime histogram between the PTU and LIF are interesting, and I think we can get more information when we access the LAS X software. But in the meantime, since the tutorial is about reading and writing files, I don't know if this is strictly related to the topic of the tutorial and may be omitted?

I think for now it is still relevant but can hopefully be removed once the photon filter is applied before exporting to PTU.

cgohlke commented on tutorials/api/phasorpy_io.py:
```diff
@@ -333,17 +334,19 @@
-# a TCSPC lifetime histograms from the
+# a phase histograms from the
 # `Convallaria <https://zenodo.org/records/14026720>`_ dataset, which was
 # acquired at the second harmonic. The dataset is a time series of two
 # channels. Since the photon count is low and the second channel empty,
```
As far as I understand, the laser frequency was 40 MHz and the FLIMbox firmware was set to encode the second harmonic histogram, so the decoded phase histogram corresponds to 80 MHz. I don't know why that option exists in the firmware. Maybe it's used just as a digital frequency multiplier? Instead of doubling the frequency, it may be cleaner to pass around both, the `frequency` and `harmonic` variables, to `phasor_calibrate`and `phasor_plot`. However, `phasor_plot` does not currently accept a `harmonic` parameter so that would need to be added to this PR.

lmalacrida replied:
I, DIVER, work with MaiTai Laser, so the frequency is 80 MHz. Phasor calculations can be performed on higher harmonics but not lower ones.  

Regarding the question above about DFD and TCSPC histograms: in the new paper (https://doi.org/10.1038/s43586-024-00358-8), we refer to both as time-tagging techniques. Both aim to detect single-photon arrivals but approach it differently. While DFD relies on a specific period, in the TCSPC approach, the only factor that matters is the time of arrival, which means there are no periodic relations. For me, this is the most fundamental difference between the two techniques. For phasor transformation, DFD is safe and period guaranteed by default. However, in TCSPC, phasor transformation may encounter issues due to decay cuts based on how the data is acquired, which can lead to poor performance in phasor calculations. We discuss this issue in our paper doi: 10.1002/jemt.23061.

cgohlke replied:
> work with MaiTai Laser, so the frequency is 80 MHz

Not according to the FBD file header.

cgohlke replied:
I will take a closer look at this file. For FBD, there's not one but three places the laser frequency may be encoded and headers may be corrupted or duplicated (what a total mess this format): 

1. file name code
2. FBD binary header
3. separate FBS XML file

cgohlke replied:
According to the (duplicate) FBD file header, the laser frequency is 40 MHz.

bruno-pannunzio commented:
Hi @cgohlke, I just received a set of FLIMLABS JSON sample files that seem to have correct harmonic information from what I could process and plot. I think this can be used for the tutorial section. Should I upload them to Zenodo on the PhasorPy community?

cgohlke commented:
Hi @bruno-pannunzio : can you please send the files to me? I would like to check them first.

cgohlke commented:
Hi @bruno-pannunzio: The new dataset looks good. I added a section about FLIM LABS JSON files to the IO tutorial and also updated the PR description with the changes. Can you take another look? Maybe @FLIMLABS has comments?

I think this PR is ready to be merged (after review) if we cannot get access to the LAS X FLIM software in the next days.

cgohlke commented:
Hi @FLIMLABS : yes, please send the JSON schema so I can compare it with assumptions I made in the implementation. As far as I understand, the JSON schema is a means to document the file format and can be used to validate files. Ideally, you publish the schema on your website or include it with your software as a reference. The schema is not directly required or used by the PhasorPy library.

bruno-pannunzio reviewed (approved):
I think this PR can be merged Christoph. I reviewed the modifications and the new additions and they are looking good and clear.

Regarding the LAS X software for comparison of the photon filter. We may have the possibility to access the software this Friday, but it depends if we can successfully connect remotely. So I suggest we merge this PR and if we have the files next week, we can create a separate PR if we need to implement some changes depending on what we find when comparing the files.

FLIMLABS commented:
Hi @cgohlke and @bruno-pannunzio please see our Json schema attached - you'll see different folders for the different type of files - I think so far phasorpy could read the "flim-imager.phasors_global.schema.json"

[data-export-json-schemas.zip](https://github.com/user-attachments/files/19211636/data-export-json-schemas.zip)

cgohlke commented:
Hi @FLIMLABS : that's great. Thank you. PhasorPy should be able to read `flim-imager.phasors_global.schema` and `flim-imager.imaging_global.schema` files. The [tutorial](https://www.phasorpy.org/docs/dev/tutorials/api/phasorpy_io/#flim-labs-json) demonstrates both, and also includes reading calibration settings from `flim.imager.calibration.schema` files.
